packages/ws-server/src/ConnectionLimiter.ts

```diff
@@ -0,0 +1,97 @@
+/*-
+ *
+ * Hedera JSON RPC Relay
+ *
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import { Logger } from "pino";
+import {WebSocketError} from "@hashgraph/json-rpc-relay";
+
+type IpCounter = {
+    [key: string]: number;
+};
+
+const IP_LIMIT_ERROR = WebSocketError.CONNECTION_IP_LIMIT_EXCEEDED;
+const TTL_EXPIRED = WebSocketError.TTL_EXPIRED;
+
+export default class ConnectionLimiter {
+    private connectedClients: number;
+    private clientIps: IpCounter;
+    private logger: Logger;
+
+    constructor(logger: Logger) {
+        this.logger = logger;
+        this.connectedClients = 0;
+        this.clientIps = {};
+    }
+
+
+    public incrementCounters(ctx) {
+        const {ip} = ctx.request;
+
+        this.connectedClients = ctx.app.server._connections;
+
+        if (!this.clientIps[ip]) {
+            this.clientIps[ip] = 1;
+        }
+        else {
+            this.clientIps[ip]++;
+        }
+        ctx.websocket.ipCounted = true;
+    }
+
+    public decrementCounters(ctx) {
+        if (ctx.websocket.ipCounted) {
+            const {ip} = ctx.request;
+            this.clientIps[ip]--;
+            if (this.clientIps[ip] === 0) delete this.clientIps[ip];
+        }
+        this.connectedClients--;
+    }
+
+    public applyLimits(ctx) {
+        const {ip} = ctx.request;
+
+        // Limit connections from a single IP address
+        const limitPerIp = parseInt(process.env.WS_CONNECTION_LIMIT_PER_IP || '10');
```

Comment:
nit: still think this should be a constant at the class level. Maybe even on instantiation of this file

Reply:
I did it this way so that it's possible to change the value after initialization - this is used in the tests. I know it's not the best solution, but the alternative is to add setter/getter methods and I think that would add unnecessary complexity.